pkg/strategy/infinity-grid/strategy.go

```diff
@@ -0,0 +1,534 @@
+
+	*bbgo.Persistence
+
+	// OrderExecutor is an interface for submitting order.
+	// This field will be injected automatically since it's a single exchange strategy.
+	bbgo.OrderExecutor `json:"-" yaml:"-"`
+
+	// Market stores the configuration of the market, for example, VolumePrecision, PricePrecision, MinLotSize... etc
+	// This field will be injected automatically since we defined the Symbol field.
+	types.Market `json:"-" yaml:"-"`
+
+	// These fields will be filled from the config file (it translates YAML to JSON)
+	Symbol string `json:"symbol" yaml:"symbol"`
+
+	LowerPrice fixedpoint.Value `json:"lowerPrice" yaml:"lowerPrice"`
+
+	// Buy-Sell Margin for each pair of orders
+	Margin fixedpoint.Value `json:"margin"`
+
+	// Quantity is the quantity you want to submit for each order.
+	Quantity fixedpoint.Value `json:"quantity"`
+
+	InitialOrderQuantity fixedpoint.Value `json:"initialOrderQuantity"`
+	CountOfMoreOrders    int              `json:"countOfMoreOrders"`
+
+	// GridNum is the grid number, how many orders you want to post on the orderbook.
+	GridNum int64 `json:"gridNumber" yaml:"gridNumber"`
+
+	// Side is the initial maker orders side. defaults to "both"
+	Side types.SideType `json:"side" yaml:"side"`
+
+	// Long means you want to hold more base asset than the quote asset.
+	Long bool `json:"long,omitempty" yaml:"long,omitempty"`
+
+	state *State
+
+	// orderStore is used to store all the created orders, so that we can filter the trades.
+	orderStore *bbgo.OrderStore
+
+	// activeOrders is the locally maintained active order book of the maker orders.
+	activeOrders *bbgo.ActiveOrderBook
+
+	tradeCollector *bbgo.TradeCollector
+
+	currentUpperGrid int
+	currentLowerGrid int
+
+	// groupID is the group ID used for the strategy instance for canceling orders
+	groupID uint32
+}
+
+func (s *Strategy) ID() string {
+	return ID
+}
+
+func (s *Strategy) Validate() error {
+	if s.LowerPrice.IsZero() {
+		return errors.New("lowerPrice can not be zero, you forgot to set?")
+	}
+
+	if s.Margin.Sign() <= 0 {
+		// If margin is empty or its value is negative
+		return fmt.Errorf("Margin should bigger than 0")
+	}
+
+	if s.Quantity.IsZero() {
+		return fmt.Errorf("Quantity can not be zero")
+	}
+
+	return nil
+}
+
+func (s *Strategy) placeInfiniteGridOrders(orderExecutor bbgo.OrderExecutor, session *bbgo.ExchangeSession) {
+	balances := session.Account.Balances()
+	log.Infof("Balances: %s", balances.String())
+	log.Infof("Base currency: %s", s.Market.BaseCurrency)   // BTC
+	log.Infof("Quote currency: %s", s.Market.QuoteCurrency) // USD
+	baseBalance, ok := balances[s.Market.BaseCurrency]
+	if !ok {
+		log.Errorf("base balance %s not found", s.Market.BaseCurrency)
+		return
+	}
+	if s.currentUpperGrid != 0 || s.currentLowerGrid != 0 {
+		// reconnect, do not place orders
+		return
+	}
+
+	quoteBalance, ok := balances[s.Market.QuoteCurrency]
+	if !ok || quoteBalance.Available <= 0 { // check available USD in balance
```

Comment:
Call .Available.IsZero(), it's not int, so you need to call method to compare.

Reply:
done